tasks/framework-tools/src/utils.js

```diff
@@ -0,0 +1,238 @@
+
+  const packageJsonLink = terminalLink(
+    'Redwood Project Path',
+    'file://' + packageJsonPath
+  )
+
+  if (!fs.existsSync(packageJsonPath)) {
+    console.log(
+      `Error: The specified ${packageJsonLink} does not have a package.json file.`
+    )
+    console.log('Expected:', packageJsonPath)
+    process.exit(1)
+  }
+
+  const packageJson = require(packageJsonPath)
+
+  // for use later, after updating packageJson's dependencies
+  function writePackageJson(packageJson) {
+    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, undefined, 2))
+  }
+
+  function updatePackageJsonDeps(dependencies) {
+    return produce(packageJson, (draftPackageJson) => {
+      if (draftPackageJson.dependencies) {
+        draftPackageJson.dependencies = {
+          ...draftPackageJson.dependencies,
+          ...dependencies,
+        }
+      } else {
+        draftPackageJson.dependencies = dependencies
+      }
+    })
+  }
+
+  return {
+    packageJson,
+    packageJsonPath,
+    packageJsonLink,
+    writePackageJson,
+    updatePackageJsonDeps,
+  }
+}
+
+const makeCopyPackageFiles =
+  (REDWOOD_PROJECT_NODE_MODULES) =>
+  ([packageName, files]) => {
+    console.log(
+      '',
+      terminalLink(
+        packageName,
+        'file://' + path.join(REDWOOD_PROJECT_NODE_MODULES, packageName)
+      ),
+      files.length,
+      'files'
+    )
+    for (const file of files) {
+      const src = path.join(
+        REDWOOD_PACKAGES_PATH,
+        packageName.replace('@redwoodjs', ''),
+        file
+      )
+      const dst = path.join(REDWOOD_PROJECT_NODE_MODULES, packageName, file)
+      fs.mkdirSync(path.dirname(dst), { recursive: true })
+      fs.copyFileSync(src, dst)
+    }
+  }
+
+function linkBinaries(REDWOOD_PROJECT_NODE_MODULES) {
+  const bins = redwoodBins()
+  for (let [binName, binPath] of Object.entries(bins)) {
+    // if the binPath doesn't exist, create it.
+    const binSymlink = path.join(REDWOOD_PROJECT_NODE_MODULES, '.bin', binName)
+    binPath = path.join(REDWOOD_PROJECT_NODE_MODULES, binPath)
+    if (!fs.existsSync(binSymlink)) {
+      fs.symlinkSync(binPath, binSymlink)
+    }
+
+    console.log('chmod +x', binName)
+    fs.chmodSync(binPath, '755')
+  }
+}
+
+function makeRegister(watcher) {
+  return function (registerObj) {
+    for (const [event, eventObject] of Object.entries(registerObj.on)) {
+      // figure out the syntax for the handler based on the event
+      let handler
+
+      switch (event) {
+        case 'ready':
+          handler = eventObject.handler
+          break
+        case 'change':
+          if (eventObject.guard) {
+            handler = (file) => {
+              if (eventObject.guard(file)) {
+                eventObject.handler(file)
+              }
+            }
+          } else {
+            handler = eventObject.handler
+          }
+      }
+
+      watcher.on(event, handler)
+    }
+  }
+}
```

Comment:
This feels a bit complicated.... I don't think it's worth it to abstract chokidar (if that's what it's doing?)

We're not using it anywhere else other than the sync function.

Reply:
When we talked about the api server file watcher, one of the things you mentioned has having a single point of entry for chokidar. So I experimented with a syntax here to see what that might look like (tried to give some context in https://github.com/redwoodjs/redwood/pull/2963#issuecomment-877299196). I know we're not using it anywhere else yet, just wanted to start working on it and get feedback for an easier use case. Happy to take it out.